Assert type consistency in the simplifier

tautschnig opened the pull request.

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -900,23 +902,56 @@ void goto_checkt::pointer_overflow_check(
     {
-      exprt overflow("overflow-"+expr.id_string(), bool_typet());
-      overflow.operands()=expr.operands();
+      exprt tmp;
```
Rename `binary_check_lhs`?

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -900,23 +902,56 @@ void goto_checkt::pointer_overflow_check(
   {
-    if(expr.operands().size()==2)
+    // The overflow checks are binary!
+    // We break these up.
```
`... e.g. for "a + b + c + d" we check "a + b", "(a + b) + c", "(a + b + c) + d"`

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -900,23 +902,56 @@ void goto_checkt::pointer_overflow_check(
+        tmp.operands().resize(i);
+      }
+
+      exprt tmp2 = expr;
```
Rename `binary_check`?

smowton commented on src/analyses/goto_check.cpp:
```diff
@@ -900,23 +902,56 @@ void goto_checkt::pointer_overflow_check(
+    exprt overflow("overflow-" + expr.id_string(), bool_typet());
+    overflow.operands() = expr.operands();
+
+    // this is overflow over integers
```
How about `// check for address space overflow`, since I guess the point is we're checking whether we wrap the VAS?

kroening commented:
Why not simply use make_binary()?

kroening commented:
Ok, that has type checks that will fail -- I would add a separate function make_pointer_arithmetic_binary, which relaxes the check, and then call that.

tautschnig commented:
I will overhaul this PR completely as I have in parts been solving the wrong problem. Notably, the expressions passed to `pointer_overflow_check` are _always_ binary. Step 1 will be adding a test that uses --pointer-overflow-check ...

tautschnig committed:
Expression simplification must not alter the type

tautschnig commented:
@kroening @smowton This PR has been almost completely changed, please take a fresh look.

tautschnig committed:
Note type inconsistencies in solver axioms and avoid simplifying

array_poolt::make_char_array_for_char_pointer generates conditional expressions
denoting possible array sizes. This results in types without a fixed size, and
is used to construct ifthenelse expressions that may have multiple types. This
is not permitted, and thus must be filtered out before passing to further
processing, such as simplification or flattening.

tautschnig commented:
This turned into a surprisingly widespread bug-fixing session. Can @romainbrenguier please take a look at the changes I had to make in the string refinement, @kroening please comment on fixes in the simplifier.

romainbrenguier commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -911,7 +911,19 @@ static optionalt<exprt> get_array(
+  while(s.id() == ID_if)
+  {
+    if_exprt s_if = to_if_expr(s);
+    simplify(s_if.cond(), ns);
```
this copies the whole `s` exprt, would it be better to just copy the condition?

romainbrenguier replied:
and why is this not part of `simplify_expr`?

tautschnig replied:
I will change this to `exprt simp_cond = simplify_expr(to_if_expr(s).cond(), ns);` with another `to_if_expr(s).{true,false}_case()` below.

romainbrenguier commented on src/util/simplify_expr_int.cpp:
```diff
@@ -894,7 +895,9 @@ bool simplify_exprt::simplify_concatenation(exprt &expr)
         bool value=op.is_true();
-        op=constant_exprt(value?ID_1:ID_0, unsignedbv_typet(1));
+        op = constant_exprt(
+          value ? ID_1 : ID_0, bitvector_typet(expr.type().id(), 1));
```
value should be const, or could be inlined here

tautschnig replied:
~True, but unrelated to the fix here? Let me know if you insist, though.~ I have made the change.

kroening commented on src/analyses/goto_check.cpp:
```diff
@@ -894,23 +894,22 @@ void goto_checkt::pointer_overflow_check(
-  }
+  if(expr.operands().size() != 2)
+    throw "pointer arithmetic expected to have exactly 2 operands";
+
```
DATA_INVARIANT?

tautschnig replied:
I'll change it right away, but I note that it is out of sync with the rest of the code in this file. But it looks like step-by-step migration is the best route forward.

smowton commented:
bump created: https://github.com/diffblue/test-gen/pull/1745

smowton commented:
Bump causes segfaults in some test-gen tests. Suggest @thk123 should allocate someone to examine why in due course.

tautschnig commented:
@smowton @thk123 I'm happy to chip in if I can be provided with a back trace (and assuming that most of the back trace is in public code).

thk123 commented on src/util/simplify_expr.cpp:
```diff
@@ -2402,6 +2416,7 @@ bool simplify_exprt::simplify_rec(exprt &expr)
 
   if(!result)
   {
+    POSTCONDITION(tmp.type() == expr.type());
```
We are hitting this invariant - will try to produce a jbmc version of the crash at some point today.

tautschnig replied:
Ok, that's  another bug in the simplifier then or some weird expression being passed in. Some of the bug fixes already in this PR are a consequence of adding this invariant.

thk123 replied:
The problem is we've simplified something that has type `pointer_typet(symbol_typet(java.lang.Object))` and end up with `pointer_typet(struct_typet(.tag = java.lang.Object)`. 

I've never really be clear whether this transformation (`symbol_typet` -> `ns.follow(symbol_type)`) should always be ignored? The bug relating to multi-dimensional arrays seems to arise from the fact the element type of the array is stored on the `symbol_typet` and is lost if it is `ns.follow`'d. Will see if happens on JBMC

tautschnig replied:
In this PR I've made a few changes to the simplifier to restore the original type when `type_before != type_after` despite `base_type_eq(type_before, type_after, ns)` (which is exactly the case for `symbol_typet` expansion). In my view, the simplifier should not change the type at all and therefore should be careful to restore the symbol type where needed. We could of course, instead, relax the `POSTCONDITION` to check for semantic equivalence (via `base_type_eq`) rather than syntactic equivalence, but then this would negatively affect examples like the multi-dimensional array that you mentioned.

So what it would take is tracking down what simplification is being performed and then fix this in the same way it's already being done several times in this PR.

thk123 replied:
I can't reproduce this yet on JBMC as failing example seems to depend on a model for `java.lang.Class`. It seems like `util/type_eq` should ignore symbolically equivalent types and could be used here, but it doesn't recurse unfortunately. Not got more time to investigate today but can have a bit more of a prod tomorrow.

tautschnig replied:
In order to debug this you will want to print `expr.id()` before this failing `POSTCONDITON` (and possibly `#define DEBUGX`). That should suffice to know which simplification rule to look at.

thk123 replied:
Found it - tautschnig/cbmc#7 - will verify fixes other issues. You might have some insight on how to create a regression that demonstrates this on the CBMC side?

tautschnig replied:
A test should look like this, except it doesn't really do the job right now, because the update to `x` is not constant-propagated.
```
#include <assert.h>

struct A
{
  struct A *next;
};

int main()
{
  struct A x, y;
  x.next = &y;
  assert(x.next == &y);
  return 0;
}
```

thk123 replied:
There appears to be more - will investigate these as well.

thk123 replied:
Ignore previous, appears to be PEBKAC. TG pointer updated - will post if passed. 

Re regression - in our previously broken example, it is simplified to a null pointer, so perhaps assigning `x.next = NULL` will work? Will quickly give it a go

thk123 replied:
Nope - works before and after using `cbmc test.c`

thk123 commented:
This is failing a (different) internal test - will see if I can fix that one in the same way

thk123 commented:
The next one is:

**`tmp.type()`**
```
  type: array
    * size: symbol
        * type: signedbv
            * width: 32
            * #c_type: signed_int
        * identifier: cprover_string_length::cprover_string_length$9!0@8#2
        * expression: symbol
            * type: signedbv
                * width: 32
                * #c_type: signed_int
            * identifier: cprover_string_length::cprover_string_length$9
            * #source_location: 
              * file: org/cprover/CProverString.java
              * line: 70
              * function: java::org.cprover.CProverString.charAt:(Ljava/lang/String;I)C
        * L0: 0
        * L1: 8
        * L2: 2
        * L1_object_identifier: cprover_string_length::cprover_string_length$9!0@8
        * #SSA_symbol: 1
    0: unsignedbv
        * width: 16
* identifier: string_refinement#char_array_member#275
```
vs.
**`expr.type()`**
```
* type: array
    * size: if
// elided - full version attached, but essentially a big load of nested if symex guard 1 && symex guard 2 etc
```

So it seems like the _type_ itself has been simplified and therefore not equal. Any insight @tautschnig?

[full_unsimplifed_expr_type.txt](https://github.com/diffblue/cbmc/files/1951700/full_unsimplifed_expr_type.txt)

tautschnig commented:
I don't think that the type has been simplified, but instead the input likely was an `ID_if` with inconsistent types - see 2cb197f in this PR. @thk123 Could you please check what the `.id()` of `expr` is/was when you're back?

thk123 commented:
@tautschnig Original expr attached
[old_expr.txt](https://github.com/diffblue/cbmc/files/1963524/old_expr.txt)

But yes you appear to be correct that the two branches of the if have different types than the if itself.

martin-cs commented:
@tautschnig Do you want this reviewing?  Am happy to do so but not sure of it's status vs. more recent patches.

tautschnig commented:
@martin-cs thanks, indeed there is work for me to get done here. I'll try to get to this as soon as possible...

martin-cs commented:
@tautschnig : At your convenience.

smowton commented:
@tautschnig do you know what needs fixing here or do you want more detail on how the problem occurs in test-gen?

tautschnig committed:
Simplifier: As we now assert type equality we can remove a safety net

If the simplifier receives an expression that isn't type-consistent it might
produce a type-inconsistent result, but we will now fail an invariant in that
case.

tautschnig commented:
This is now ready for review - @romainbrenguier in particular, but maybe also @martin-cs and @peterschrammel.

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -869,7 +912,10 @@ void string_refinementt::add_lemma(
   exprt simple_lemma = lemma;
   if(simplify_lemma)
+  {
+    simple_lemma = adjust_if_recursive(std::move(simple_lemma), ns);
```
This will probably not get rid of all the inconsistencies, so I'm not sure about how useful it is. Ideally we should get rid of the introduction of type inconsistencies, but I'm not sure about the best way to do that.

tautschnig replied:
I recall having tried removing them completely (which would be done by using type casts), but that did not seem to work very well. It's the long-term solution, and maybe that's what we should actually invest in instead of this hack. Not sure though, I'm certainly curious about the results with TG!

romainbrenguier replied:
Everything is green on our side, so nothing blocking merging it.

romainbrenguier reviewed (commented):
I will do a TG pull request to check no invariant fails on our tests.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 861658616de5d16a50ffd58d622a20f7cb630f87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103493328
